cmd/runtimetest/main.go

```diff
@@ -278,63 +354,148 @@ func validateRlimits(spec *rspec.Spec, t *tap.T) error {
 		}
-		if rlimit.Max != r.Hard {
-			return specerror.NewError(specerror.PosixProcRlimitsHardMatchMax, fmt.Errorf("%v rlimit hard expected: %v, actual: %v", r.Type, r.Hard, rlimit.Max), rspec.Version)
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"type":      r.Type,
+			"expected":  r.Soft,
+			"actual":    rlimit.Cur,
+		})
+
+		rfcError, err = c.Ok(rlimit.Max == r.Hard, specerror.PosixProcRlimitsHardMatchMax, spec.Version, fmt.Sprintf("has expected hard %v", r.Type))
+		if err != nil {
+			return err
 		}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"type":      r.Type,
+			"expected":  r.Hard,
+			"actual":    rlimit.Max,
+		})
 	}
 	return nil
 }
 
-func validateSysctls(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateSysctls(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.Sysctl == nil {
+		c.harness.Skip(1, "linux.sysctl not set")
 		return nil
 	}
+
 	for k, v := range spec.Linux.Sysctl {
 		keyPath := filepath.Join("/proc/sys", strings.Replace(k, ".", "/", -1))
 		vBytes, err := ioutil.ReadFile(keyPath)
 		if err != nil {
 			return err
 		}
 		value := strings.TrimSpace(string(bytes.Trim(vBytes, "\x00")))
-		if value != v {
-			return fmt.Errorf("Sysctl %v value expected: %v, actual: %v", k, v, value)
-		}
+		c.harness.Ok(value == v, fmt.Sprintf("has expected sysctl %v", k))
+		c.harness.YAML(map[string]string{
+			"sysctl":   k,
+			"expected": v,
+			"actual":   value,
+		})
 	}
 	return nil
 }
 
-func testWriteAccess(path string) error {
-	tmpfile, err := ioutil.TempFile(path, "Test")
+func testReadAccess(path string) (readable bool, err error) {
+	fi, err := os.Stat(path)
 	if err != nil {
-		return err
+		return false, err
 	}
+	if fi.Mode()&os.ModeType == 0 {
+		return testFileReadAccess(path)
+	}
+	return false, fmt.Errorf("cannot test read access for %q (mode %d)", path, fi.Mode())
+}
+
+func testFileReadAccess(path string) (readable bool, err error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return false, nil
+	}
+	defer f.Close()
+	b := make([]byte, 1)
+	_, err = f.Read(b)
+	if err == nil {
+		return true, nil
+	} else if err == io.EOF {
+		return false, nil
+	}
+	return false, err
+}
 
+func testWriteAccess(path string) (writable bool, err error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return false, err
+	}
+	if fi.IsDir() {
+		return testDirectoryWriteAccess(path)
+	} else if fi.Mode()&os.ModeType == 0 {
+		return testFileWriteAccess(path)
+	}
+	return false, fmt.Errorf("cannot test write access for %q (mode %d)", path, fi.Mode())
+}
+
+func testDirectoryWriteAccess(path string) (writable bool, err error) {
+	tmpfile, err := ioutil.TempFile(path, "Test")
+	if err != nil {
+		return false, nil
+	}
 	tmpfile.Close()
-	os.RemoveAll(filepath.Join(path, tmpfile.Name()))
+	if err != nil {
```

Comment:
this branch cannot be taken because it was tested a few lines above.

Reply:
> this branch cannot be taken because it was tested a few lines above.

Thanks :).  I've dropped it with d7e4d91 → 857249d.

Reply:
Can it use [WriteFile](https://golang.org/pkg/io/ioutil/#WriteFile)?

Reply:
> Can it use WriteFile?

Good idea.  I've updated to use `WriteFile` for the whole PR with 7ecd679 → 92de75c.